compiler/dialects/accfg.md

```diff
@@ -0,0 +1,141 @@
+This makes reasoning simple and effective.
+
+The deduplication pattern recursively looks at the operations producing the input state to determine if a value
+has already been set. If it sees that a field is set to the exact same SSA value again, it removes that setup. This
+requires the input IR to be in a canonical form where identical operation sequences have been deduplicated. This will
+miss some cases, but is an easy over-approximation, as we know that SSA values will never be re-assigned.
+
+The last two rewrites focus on moving setup operations around so that they can be optimised by the previous three
+patterns. (4) moves setup operations upwards into conditional (given that there are other setup operations inside
+the conditional). The move is only considered legal if there are no launch operations between the setup and the
+conditional.
+
+The final rewrite (5) inspects a `setup` operation inside an `scf.for` loop. If it finds at least one field that is
+set ot a loop-independent value, it creates a new setup operation before the loop that sets the loop invariant fields.
+Pattern (3) then takes care of erasing the redundant fields inside the for loop.
+
+
+### `accfg-setup-overlap`
+
+This pass tries to move setup operations upwards *before* `await` operations. This assumes a few things:
+
+- We can freely modify the accelerator state while it's running
+- We have a truly asynchronous accelerator (e.g. `launch` will not block until the accelerator has completed)
+- The accelerator dominate the runtime, meaning we can do some work while we wait (e.g. calculating the next setup values)
+
+In detail, it performs the following optimisations:
+
+1. Block level await overlap
+2. Loop level await overlap
+
+Block level await overlap tries to find a `launch() -> await() -> setup()` triplet inside the same block operating on
+the same accelerator. It then moves the `setup()` upwards until it happens between the `launch` and the `await`
+resulting in `launch() -> setup() -> await()`. The operations producing the values that are set up are moved together
+with the setup, as long as they don't have any effects (e.g. memory writes). If any of the operations that produce the
+inputs cannot be determined to be effect free, the move is aborted.
+
+The loop level await overlap peels the setup operation one loop iteration "upwards", resulting in a setup *before* the
+loop operating on the initial loop counters, together with a replaced version at the end of the loop body that operates
+on the loop counter incremented by the loop step. This effectively moves only the setup operation "upwards" inside the
+loop. The resulting IR contains one more `setup` call than the original version (as we now also set up the accelerator
+to the `ub+1` state), but since this state is never launched, no adverse effects are expected. This pattern effectively
```

Comment:
`ub+1` reads like undefined behaviour + 1. Also should this not be `iv` instead?